File: labs/control-ui/api-shadow/reverse_tunnel_relay.ts
export interface UiReverseSession {
  sessionId: number;
  remoteAddr: string;
  targetAddr: string;
  bytesIn: number;
  bytesOut: number;
  isConnected: boolean;
}

export class ReverseTunnelRelay {
  private sessions: Map<number, UiReverseSession> = new Map();
  private nextId = 1;
  private maxSessions: number;
  constructor(maxSessions: number = 100) {
    this.maxSessions = maxSessions;
  }

  openSession(remoteAddr: string, targetAddr: string): number | null {
    if (this.sessions.size >= this.maxSessions) return null;
    const id = this.nextId++;
    this.sessions.set(id, {
      sessionId: id,
      remoteAddr,
      targetAddr,
      bytesIn: 0,
      bytesOut: 0,
      isConnected: true,
    });
    return id;
  }

  recordTraffic(sessionId: number, inBytes: number, outBytes: number): boolean {
    const s = this.sessions.get(sessionId);
    if (!s || !s.isConnected) return false;
    s.bytesIn += inBytes;
    s.bytesOut += outBytes;
    return true;
  }

  closeSession(sessionId: number): boolean {
    const s = this.sessions.get(sessionId);
    if (!s) return false;
    s.isConnected = false;
    this.sessions.delete(sessionId);
    return true;
  }

  activeSessions(): number {
    return this.sessions.size;
  }
}

File: labs/mobile/donor-archive/tunnel/ReverseTunnelRelay.kt
package com.luminet.android.tunnel

data class AndroidReverseSession(
    val sessionId: Long,
    val remoteAddr: String,
    val targetAddr: String,
    var bytesIn: Long = 0,
    var bytesOut: Long = 0,
    var isConnected: Boolean = true
)

class ReverseTunnelRelay(private val maxSessions: Int = 100) {
    private val sessions = mutableMapOf<Long, AndroidReverseSession>()
    private var nextId = 1L

    @Synchronized
    fun openSession(remoteAddr: String, targetAddr: String): Long? {
        if (sessions.size >= maxSessions) return null
        val id = nextId++
        sessions[id] = AndroidReverseSession(id, remoteAddr, targetAddr)
        return id
    }

    @Synchronized
    fun recordTraffic(sessionId: Long, inBytes: Long, outBytes: Long): Boolean {
        val s = sessions[sessionId] ?: return false
        if (!s.isConnected) return false
        s.bytesIn += inBytes
        s.bytesOut += outBytes
        return true
    }

    @Synchronized
    fun closeSession(sessionId: Long): Boolean {
        val s = sessions.remove(sessionId) ?: return false
        s.isConnected = false
        return true
    }

    fun activeSessions(): Int = sessions.size
}
